File: cookbook/php/Imports/createAutoImport.php
<?php

//Url de base
$urlBase = 'http://v8.mailperformance.com/';

//X-Key
$xKey = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';



function getCreateImportJson()
{
    // Remplissez les informations obligatoires
    $importName = "Nom de votre Import automatique"; //Nom de l'import
    $schedulerName = "Nom du scheduler"; //Nom du scheduler
    $binding = 1234; // Id du binding
    $segmentId = 1234; // Id du segment
    $fieldsId = 1234; // Id du field
    $contactsId = array("1234ABCD"); // Id des utilisateurs : Administration -> Utilisateurs -> Identifiant
    $groupsContactsId = array(1234); // Id des groupes : Administration -> Groupes -> Identifiant



    $createImportData = array(
        "name" => $importName,
        "binding" => $binding,
        "source" => array(
                "type" =>"ftp",
        ),
        "scheduler" => array(
            "type" => "periodic",
            "name" => $schedulerName,
            "frequency" => array(
                "occurs" => array(
                    "type" => "daily", // Quotidien

                    //Autres possibilites
                    /*
                    "type" => "weekly", // Hebdomadaire
                    "days" => array("Mon", "Tue", "Wed", "Thu", "Fri", "Sat"), // les jours selectionnes de la semaine

                    "type" => "monthly", // Mensuel
                    "days" => array( 20 ) // le numeros du jour des imports
                    */
                ),
                "periodicity" => array( // Heure de l'import
                    "type" => "once",
                    "value" => array(
                        "hour" => 6,
                        "minute" => 0,
                        "second" => 0
                    )
                )
            ),
            "validity" => array(  // La validation n'est pas obligatoire
                "start" => array( // Date du commencement
                    "year" => 2016, // Les dates doivent etre coherentes
                    "month" => 12,
                    "date" => 1,
                    "hour" => 0,
                    "minute" => 0,
                    "second" => 0
                ),
                "end" => array( // Date de fin
                    "year" => 2017, // Les dates doivent etre coherentes
                    "month" => 10,
                    "date" => 1,
                    "hour" => 0,
                    "minute" => 0,
                    "second" => 0
                )
            )
        ),
        "features" => array( // POUR SUPPRIMER NIMPORTE QUELLE FONCTION IL SUFFIT DE METTRE EN COMMENTAIRE !!! (exemple avec 'redList')
            array( // Parametres de normalisation
                "type" => "normalization",
                "fields" => array(
                    $fieldsId => array("FRA") // Pour changer de pays de normalisation utiliser l'ISO ALPHA-3 Code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3)
                )
            ),
            array( // Ajout des cibles a un Segment
                "type" => "segmentation",
                "segmentId" => $segmentId,
                "emptyExisitingSegment" => true
            ),/*
            array( // Ajout de TOUTES les cibles (de l'import) a la liste rouge
                "type" => "redList",
                "destination" => array( "sms" => true, "email" => true )
            ),*/
            array( // Regles a appliquer en cas de doublon
                "type" => "duplicate",
                "rules" => array( "ignore" => true ) /// ou "rules" => array( "first" => true ) ou "rules" => array( "last" => true )
            ),
            array( // Parametres des rapports
                "type" => "report",
                "sendFinalReport" => false, //Envoi d'un rapport final
                "sendErrorReport" => true, //Envoi d'un rapport d'erreur

                //Si un des rapport doit être envoyer, il faut au moins un destinataire ('contactGuids'/'groupIds')
                "contactGuids" => $contactsId, // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifiant
                "groupIds" => $groupsContactsId, // Id des groupes : Administration -> Groupes -> Idenfitifiant
            ),
            array( // Parametres pour mettre a jour une cible dans la base de donnees
                "type" => "database",
                "updateExisting" => true,
                "crushData" => true
            )
        )
    );

    $createImportJson = json_encode($createImportData);

    //on affiche le json
    echo "json :\n" . $createImportJson . "\n\n";

    return ($createImportJson);
}



//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DES CREATIONs DES JSONS//////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DU PROGRAMME///////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


$req = NULL;

//Nouvelle url
$url = $urlBase . "imports/";

//lancement de la requete
$con = urlPostOrPut($req, 'POST', getCreateImportJson(), $xKey, $url);
if (($req = verification($con)) == null)
    return (0);

curl_close($req);
return (1);



//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DU PROGRAMME/////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DES FONCTIONS//////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

//On verifie les reponses
function verification($con)
{
    $result = $con['result'];
    $info = $con['info'];
    $req = $con['req'];

    //Verification des reponses
    if ($result == false)
    {
        //Erreur lors de la requete

        //Affichage de l'erreur
        $info = curl_getinfo($req);
        echo 'Error  => ' . $info['http_code'] . "\n";
        curl_close($req);
        return (null);
    }
    else
    {
        //On affiche la reponse
        echo 'OK : ' . $result . "\n\n";
        return ($req);
    }
}



//Utilisation de cURL pour remplir les requetes
//Fonctions de connexion
function startCurlInit($url)
{
    $init = curl_init();
    curl_setopt($init, CURLOPT_URL, $url);
    curl_setopt($init, CURLOPT_RETURNTRANSFER, true);
    return ($init);
}

function urlPostOrPut($req, $request, $dataJson, $xKey, $url)
{
    //On remplit la requete avec le bon verbe ($request) : POST / PUT
    $req = startCurlInit($url);
    curl_setopt($req, CURLOPT_CUSTOMREQUEST, $request);
    curl_setopt($req, CURLOPT_POSTFIELDS, $dataJson);

    //Mise en place du xKey et des options
    curl_setopt($req, CURLOPT_HTTPHEADER, array(
        'X-Key: ' . $xKey,
        'Content-Type: application/json',
        'Accept: application/vnd.mperf.v8.import.v1+json',
        'Content-Length: ' . strlen($dataJson),
    ));

    //Execution de la requete
    $result = curl_exec($req);

    //Verification des reponses
    $info = curl_getinfo($req);

    return (array('result' => $result, 'info' => $info, 'req' => $req));
}

?>

File: java/Imports/createAutoImport.java
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class createAutoImport
{
    public static JSONObject getCreateImportJson() throws JSONException
    {
        // Remplissez les informations obligatoires
        String importName = "Nom de votre Import automatique"; // Nom de l'import
        String schedulerName = "Nom du scheduler"; // Nom du scheduler
        int binding = 1234; // Id du binding
        int segmentId = 1234; // Id du segment
        int fieldsId = 1234; // Id du field
        String contactsId[] = {"1234ABCD"}; // Id des utilisateurs : Administration -> Utilisateurs -> Identifiant
        int groupsContactsId[] = {1234}; // Id des groupes : Administration -> Groupes -> Identifiant

        //Creation du Json


        JSONObject source = new JSONObject();
        source.put("type", "ftp");

        JSONObject occurs = new JSONObject();
        occurs.put("type", "daily"); //Quotidient
        /* Autres possibilites :
        occurs.put("type", "weekly"); // Hebdomadaire
        String days[] = {"Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
        occurs.put("days", days);

        occurs.put("type", "monthly"); // Mensuel
        int days[] = {20};
        occurs.put("days", days);
        */

        JSONObject periodicityValue = new JSONObject();
        periodicityValue.put("hour", 6);
        periodicityValue.put("minute", 0);
        periodicityValue.put("second", 0);

        JSONObject periodicity = new JSONObject(); // Heure de l'import
        periodicity.put("type", "once");
        periodicity.put("value", periodicityValue);

        JSONObject frequency = new JSONObject();
        frequency.put("occurs", occurs);
        frequency.put("periodicity", periodicity);

        JSONObject validityStart = new JSONObject(); // Les dates doivent etre coherentes avec la date de creation
        validityStart.put("year", 2016);
        validityStart.put("month", 12);
        validityStart.put("date", 1);
        validityStart.put("hour", 0);
        validityStart.put("minute", 0);
        validityStart.put("second", 0);

        JSONObject validityEnd = new JSONObject(); // Les dates doivent etre coherentes avec la date de creation
        validityEnd.put("year", 2017);
        validityEnd.put("month", 10);
        validityEnd.put("date", 1);
        validityEnd.put("hour", 0);
        validityEnd.put("minute", 0);
        validityEnd.put("second", 0);

        JSONObject validity = new JSONObject(); // La validation n'est pas obligatoire
        validity.put("start", validityStart);
        validity.put("end", validityEnd);

        JSONObject scheduler = new JSONObject();
        scheduler.put("type", "periodic");
        scheduler.put("name", schedulerName);
        scheduler.put("frequency", frequency);
        scheduler.put("validity", validity);

        JSONObject fields = new JSONObject();
        String country[] = {"FRA"}; // Pour changer de pays de normalisation utiliser l'ISO ALPHA-3 Code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3)
        fields.put(Integer.toString(fieldsId), country);

        JSONObject normalization = new JSONObject(); // Parametre de normalisation
        normalization.put("type", "normalization");
        normalization.put("fields", fields);

        JSONObject segmentation = new JSONObject(); // Ajout des cibles a un Segment
        segmentation.put("type", "segmentation");
        segmentation.put("segmentId", segmentId);
        segmentation.put("emptyExisitingSegment", true);

        JSONObject destination = new JSONObject();
        destination.put("sms", true);
        destination.put("email", true);

        JSONObject redList = new JSONObject(); // Ajout des cibles a la liste rouge
        redList.put("type", "redList");
        redList.put("destination", destination);

        JSONObject rules = new JSONObject();
        rules.put("ignore", true); //ou rules.put("first", true); ou rules.put("last", true);

        JSONObject duplicate = new JSONObject(); // Regles a appliquer en cas de doublon
        duplicate.put("type", "duplicate");
        duplicate.put("rules", rules);

        JSONObject report = new JSONObject(); // Parametres du rapport
        report.put("type", "report");
        report.put("sendFinalReport", false);
        report.put("sendErrorReport", true);
        // Si un des rapport doit être envoyer, il faut au moins un destinataire ('contactGuids'/'groupIds')
        report.put("contactGuids", contactsId); // Id des utilisateurs : Administration -> Utilisateurs -> Identifiant
        report.put("groupIds", groupsContactsId); // Id des groupes : Administration -> Groupes -> Identifiant

        JSONObject database = new JSONObject(); // Parametres pour mettre a jour une cible dans la base de donnees
        database.put("type", "database");
        database.put("updateExisting", true);
        database.put("crushData", true);

        JSONArray features = new JSONArray(); // Si vous voulez enlever une option il suffit de mettre la ligne en commentaire
        features.put(normalization);
        features.put(segmentation);
        //features.put(redList);
        features.put(duplicate);
        features.put(report);
        features.put(database);

        JSONObject createImportJson = new JSONObject();
        createImportJson.put("name", importName);
        createImportJson.put("binding", binding);
        createImportJson.put("source", source);
        createImportJson.put("scheduler", scheduler);
        createImportJson.put("features", features);

        //On affiche le Json
        System.out.print("Json :\n" + createImportJson + "\n\n");

        return (createImportJson);
    }


//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DE CREATION DES JSONS////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DU PROGRAMME///////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


    public static void main(String[] args) throws Exception
    {
        //Url de base
        String urlBase = "https://backoffice.mailperformance.com/";
        //X-Key
        String xKey = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

        //Lancement de la connexion pour remplir la requete
        String url = urlBase + "imports/";
        HttpURLConnection con = null;

        System.out.print("CREATE IMPORT\n");

        con = urlPostOrPut(con, "POST", getCreateImportJson(), xKey, url);

        //Lecture des donnees
        BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
        String reply = buffRead.readLine();
        buffRead.close();
        System.out.print(reply);

        con.disconnect();
        System.exit(1);
    }


/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DU PROGRAMME////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DES FONCTIONS/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


    //Fonctions de connexion
    public static HttpURLConnection startCon(String url, String xKey) throws MalformedURLException, IOException
    {
        //Lancement de la connexion
        HttpURLConnection con = (HttpURLConnection)new URL(url).openConnection();

        //Mise en place du xKey et des options
        con.setRequestProperty("X-Key", xKey);
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("Accept", "application/vnd.mperf.v8.import.v1+json");
        return (con);
    }

    public static HttpURLConnection urlPostOrPut(HttpURLConnection con, String request, JSONObject dataJson, String xKey, String url) throws MalformedURLException, IOException
    {
        //Lancement de la connexion pour remplir la requete
        con = startCon(url, xKey);
        con.setRequestProperty("Content-Length", Integer.toString(dataJson.toString().length()));
        con.setRequestMethod(request);
        con.setDoOutput(true);

        //Envoie des informations dans la connexion
        DataOutputStream streamCon = new DataOutputStream(con.getOutputStream());
        streamCon.write(dataJson.toString().getBytes());
        streamCon.flush();
        streamCon.close();

        //Verification des reponses
        int responseCode = con.getResponseCode();
        if (responseCode != 200 && responseCode != 204)
        {
            //Affichage de l'erreur
            System.out.print("Error : " + responseCode + ' ' + con.getResponseMessage() + ".\nEnd urlPostOrPut");
            System.exit(1);
        }
        return (con);
    }
}
